# CS6320 Assignment 1 - Opinion Spam Classification

## Madhumita Ramesh
## Akshaiy Ramlinkam Jayakumar Srilatha

<br>

<br>

## Training data (Truthful)

In [1]:
import pandas as pd

In [3]:
train_truthful = pd.read_csv("./A1_DATASET/train/truthful.txt", sep="\n", header=None)

In [39]:
train_truthful.head()

0
0  I booked two rooms four months in advance at t...
1  I LOVED this hotel . The room was so chic and ...
2  The Hard Rock Hotel Chicago has become my favo...
3  We also just returned from a frigid weekend in...
4  I can easily say this is one of the worst hote...

In [60]:
len(train_truthful)

512

In [40]:
train_truthful[0][511]

'We booked our hotel stay thru Yahoo and requested a room for 4 with 2 beds . When we got to the hotel , not only did they not have a room for us ( we pre-paid ! ) , but they said they could not guarantee what we had requested online , they could only give us what was available . The hotel staff was rude and condescending and only after I got upset did they finally tell me they did have a room for us . If you book online - beware ! You may not get what you paid for . They did not stand by their promise they advertise in each room .'

## Building the unigram dictionary (Truthful)

In [117]:
train_truthful_dictionary = {}
for row in range(len(train_truthful)):
    for word in train_truthful[0][row].split():
        if word.isalpha():
            if word not in train_truthful_dictionary:
                train_truthful_dictionary[word]=1
            else:
                train_truthful_dictionary[word]+=1

In [291]:
train_truthful_dictionary

{'I': 1659,
 'booked': 81,
 'two': 120,
 'rooms': 183,
 'four': 11,
 'months': 7,
 'in': 1212,
 'advance': 7,
 'at': 717,
 'the': 4250,
 'Talbott': 28,
 'We': 461,
 'were': 578,
 'placed': 8,
 'on': 610,
 'top': 38,
 'floor': 135,
 'next': 99,
 'to': 2077,
 'elevators': 31,
 'which': 173,
 'are': 308,
 'used': 31,
 'all': 245,
 'night': 198,
 'long': 37,
 'When': 88,
 'speaking': 3,
 'front': 121,
 'desk': 155,
 'was': 1820,
 'told': 77,
 'that': 662,
 'they': 380,
 'simply': 17,
 'honoring': 1,
 'my': 419,
 'request': 21,
 'for': 997,
 'an': 219,
 'upper': 5,
 'had': 525,
 'requested': 32,
 'a': 2215,
 'better': 89,
 'view': 101,
 'am': 62,
 'looking': 43,
 'brick': 6,
 'wall': 30,
 'and': 2552,
 'getting': 28,
 'no': 212,
 'sleep': 28,
 'He': 21,
 'also': 105,
 'me': 242,
 'received': 19,
 'complaints': 12,
 'before': 61,
 'from': 348,
 'guests': 52,
 'aware': 9,
 'of': 1040,
 'noise': 29,
 'problem': 38,
 'Why': 6,
 'then': 59,
 'did': 242,
 'place': 85,
 'us': 212,
 'this': 453,
 '

In [119]:
total_word_count = sum(train_truthful_dictionary.values())
print(total_word_count)

77148


In [120]:
train_truthful_dictionary_unigram = {key:val/total_word_count for key,val in train_truthful_dictionary.items()}

In [121]:
train_truthful_dictionary_unigram

{'I': 0.021504121947425726,
 'booked': 0.0010499300046663555,
 'two': 0.0015554518587649713,
 'rooms': 0.0023720640846165812,
 'four': 0.0001425830870534557,
 'months': 9.073469176128999e-05,
 'in': 0.015710063773526208,
 'advance': 9.073469176128999e-05,
 'at': 0.009293824856120703,
 'the': 0.05508891999792606,
 'Talbott': 0.00036293876704515996,
 'We': 0.005975527557422098,
 'were': 0.007492093119717945,
 'placed': 0.00010369679058433142,
 'on': 0.00790688028205527,
 'top': 0.0004925597552755742,
 'floor': 0.0017498833411105927,
 'next': 0.0012832477834811013,
 'to': 0.026922279255457045,
 'elevators': 0.00040182506351428424,
 'which': 0.002242443096386167,
 'are': 0.003992326437496759,
 'used': 0.00040182506351428424,
 'all': 0.0031757142116451494,
 'night': 0.0025664955669622027,
 'long': 0.0004795976564525328,
 'When': 0.0011406646964276456,
 'speaking': 3.888629646912428e-05,
 'front': 0.0015684139575880126,
 'desk': 0.002009125317571421,
 'was': 0.023591019857935398,
 'told': 0.

## Building the bigram dictionary (Truthful)

In [128]:
train_truthful_dictionary_pairs = {}
for row in range(len(train_truthful)):
    row_as_list = train_truthful[0][row].split()
    row_as_list.insert(0,"START")
    row_as_list.append("END")
    row_as_list = [values for values in row_as_list if values.isalpha()]
    
    for i in range(len(row_as_list)-1):
        if row_as_list[i].isalpha():
            if (row_as_list[i],row_as_list[i+1]) not in train_truthful_dictionary_pairs.keys():
                train_truthful_dictionary_pairs[(row_as_list[i],row_as_list[i+1])]=1
            else:
                train_truthful_dictionary_pairs[(row_as_list[i],row_as_list[i+1])]+=1

In [129]:
train_truthful_dictionary_pairs

{('START', 'I'): 109,
 ('I', 'booked'): 20,
 ('booked', 'two'): 1,
 ('two', 'rooms'): 3,
 ('rooms', 'four'): 1,
 ('four', 'months'): 1,
 ('months', 'in'): 2,
 ('in', 'advance'): 7,
 ('advance', 'at'): 1,
 ('at', 'the'): 329,
 ('the', 'Talbott'): 22,
 ('Talbott', 'We'): 2,
 ('We', 'were'): 62,
 ('were', 'placed'): 1,
 ('placed', 'on'): 2,
 ('on', 'the'): 213,
 ('the', 'top'): 11,
 ('top', 'floor'): 4,
 ('floor', 'next'): 1,
 ('next', 'to'): 25,
 ('to', 'the'): 263,
 ('the', 'elevators'): 10,
 ('elevators', 'which'): 2,
 ('which', 'are'): 3,
 ('are', 'used'): 1,
 ('used', 'all'): 1,
 ('all', 'night'): 13,
 ('night', 'long'): 4,
 ('long', 'When'): 1,
 ('When', 'speaking'): 1,
 ('speaking', 'to'): 1,
 ('the', 'front'): 80,
 ('front', 'desk'): 97,
 ('desk', 'I'): 4,
 ('I', 'was'): 183,
 ('was', 'told'): 20,
 ('told', 'that'): 19,
 ('that', 'they'): 35,
 ('they', 'were'): 39,
 ('were', 'simply'): 1,
 ('simply', 'honoring'): 1,
 ('honoring', 'my'): 1,
 ('my', 'request'): 3,
 ('request', 'for'

In [125]:
total_bigram_count = sum(train_truthful_dictionary_pairs.values())
print(total_bigram_count)

77660


In [130]:
train_truthful_dictionary_bigram = {}
for key,val in train_truthful_dictionary_pairs.items():
    if key[0] == "START":
        train_truthful_dictionary_bigram[key] = val/len(train_truthful)
    else:
        train_truthful_dictionary_bigram[key] = val/train_truthful_dictionary[key[0]]

In [131]:
train_truthful_dictionary_bigram

{('START', 'I'): 0.212890625,
 ('I', 'booked'): 0.012055455093429777,
 ('booked', 'two'): 0.012345679012345678,
 ('two', 'rooms'): 0.025,
 ('rooms', 'four'): 0.00546448087431694,
 ('four', 'months'): 0.09090909090909091,
 ('months', 'in'): 0.2857142857142857,
 ('in', 'advance'): 0.005775577557755775,
 ('advance', 'at'): 0.14285714285714285,
 ('at', 'the'): 0.45885634588563456,
 ('the', 'Talbott'): 0.0051764705882352945,
 ('Talbott', 'We'): 0.07142857142857142,
 ('We', 'were'): 0.13449023861171366,
 ('were', 'placed'): 0.0017301038062283738,
 ('placed', 'on'): 0.25,
 ('on', 'the'): 0.34918032786885245,
 ('the', 'top'): 0.0025882352941176473,
 ('top', 'floor'): 0.10526315789473684,
 ('floor', 'next'): 0.007407407407407408,
 ('next', 'to'): 0.25252525252525254,
 ('to', 'the'): 0.1266249398170438,
 ('the', 'elevators'): 0.002352941176470588,
 ('elevators', 'which'): 0.06451612903225806,
 ('which', 'are'): 0.017341040462427744,
 ('are', 'used'): 0.003246753246753247,
 ('used', 'all'): 0.032

## Training data (Deceptive)

In [58]:
import pandas as pd

In [ ]:
train_deceptive = pd.read_csv("./A1_DATASET/train/deceptive.txt", sep="\n", header=None)

In [59]:
train_deceptive.head()

0
0  I was here on business so I needed to get some...
1  My wife and I were excited to visit and shop a...
2  I had an awful experience here . The staff wer...
3  I stayed at the Affnia hotel in Chicago and wa...
4  My stay at the Omni Chicago Hotel was wonderfu...

## Building the unigram dictionary (Deceptive)

In [135]:
train_deceptive_dictionary = {}
for row in range(len(train_deceptive)):
    for word in train_deceptive[0][row].split():
        if word.isalpha():
            if word not in train_deceptive_dictionary:
                train_deceptive_dictionary[word]=1
            else:
                train_deceptive_dictionary[word]+=1

In [136]:
train_deceptive_dictionary

{'I': 2485,
 'was': 2051,
 'here': 105,
 'on': 421,
 'business': 118,
 'so': 230,
 'needed': 31,
 'to': 2310,
 'get': 176,
 'some': 101,
 'rest': 19,
 'the': 4089,
 'next': 118,
 'day': 101,
 'but': 416,
 'walls': 24,
 'were': 573,
 'thin': 13,
 'that': 654,
 'could': 202,
 'all': 284,
 'carrying': 2,
 'in': 1233,
 'neighbors': 8,
 'room': 902,
 'till': 2,
 'am': 49,
 'This': 117,
 'very': 340,
 'beautiful': 76,
 'building': 16,
 'a': 1846,
 'lot': 28,
 'smaller': 16,
 'than': 120,
 'expected': 40,
 'and': 2522,
 'staff': 247,
 'not': 522,
 'friendly': 90,
 'The': 1031,
 'person': 30,
 'at': 805,
 'front': 122,
 'desk': 163,
 'had': 600,
 'no': 135,
 'personality': 2,
 'did': 260,
 'even': 136,
 'welcome': 14,
 'me': 281,
 'hotel': 892,
 'All': 43,
 'do': 132,
 'think': 40,
 'will': 184,
 'be': 393,
 'back': 138,
 'My': 187,
 'wife': 84,
 'excited': 11,
 'visit': 71,
 'shop': 5,
 'Chicago': 661,
 'Magnificent': 15,
 'Mile': 17,
 'Michigan': 19,
 'We': 265,
 'chose': 16,
 'Intercontinen

In [137]:
total_word_count = sum(train_deceptive_dictionary.values())
print(total_word_count)

75582


In [138]:
train_deceptive_dictionary_unigram = {key:val/total_word_count for key,val in train_deceptive_dictionary.items()}

In [139]:
train_deceptive_dictionary_unigram

{'I': 0.032878198512873436,
 'was': 0.027136090603582862,
 'here': 0.0013892196554735254,
 'on': 0.005570109285279564,
 'business': 0.0015612182794845335,
 'so': 0.003043052578656294,
 'needed': 0.00041015056494932655,
 'to': 0.03056283242041756,
 'get': 0.002328596755841338,
 'some': 0.0013362970019316768,
 'rest': 0.00025138260432378077,
 'the': 0.05410018258315472,
 'next': 0.0015612182794845335,
 'day': 0.0013362970019316768,
 'but': 0.005503955968352253,
 'walls': 0.00031753592125109155,
 'were': 0.00758117011986981,
 'thin': 0.0001719986240110079,
 'that': 0.008652853854092243,
 'could': 0.0026725940038633536,
 'all': 0.0037575084014712496,
 'carrying': 2.6461326770924294e-05,
 'in': 0.016313407954274828,
 'neighbors': 0.00010584530708369718,
 'room': 0.011934058373686857,
 'till': 2.6461326770924294e-05,
 'am': 0.0006483025058876452,
 'This': 0.0015479876160990713,
 'very': 0.00449842555105713,
 'beautiful': 0.001005530417295123,
 'building': 0.00021169061416739435,
 'a': 0.0244

## Building the bigram dictionary (Deceptive)

In [140]:
train_deceptive_dictionary_pairs = {}
for row in range(len(train_deceptive)):
    row_as_list = train_deceptive[0][row].split()
    row_as_list.insert(0,"START")
    row_as_list.append("END")
    row_as_list = [values for values in row_as_list if values.isalpha()]
    
    for i in range(len(row_as_list)-1):
        if row_as_list[i].isalpha():
            if (row_as_list[i],row_as_list[i+1]) not in train_deceptive_dictionary_pairs.keys():
                train_deceptive_dictionary_pairs[(row_as_list[i],row_as_list[i+1])]=1
            else:
                train_deceptive_dictionary_pairs[(row_as_list[i],row_as_list[i+1])]+=1

In [337]:
train_deceptive_dictionary_pairs

{('START', 'I'): 134,
 ('I', 'was'): 325,
 ('was', 'here'): 3,
 ('here', 'on'): 3,
 ('on', 'business'): 14,
 ('business', 'so'): 4,
 ('so', 'I'): 35,
 ('I', 'needed'): 14,
 ('needed', 'to'): 10,
 ('to', 'get'): 74,
 ('get', 'some'): 5,
 ('some', 'rest'): 1,
 ('rest', 'the'): 1,
 ('the', 'next'): 41,
 ('next', 'day'): 11,
 ('day', 'but'): 4,
 ('but', 'the'): 50,
 ('the', 'walls'): 14,
 ('walls', 'were'): 6,
 ('were', 'so'): 18,
 ('so', 'thin'): 3,
 ('thin', 'that'): 1,
 ('that', 'I'): 84,
 ('I', 'could'): 73,
 ('could', 'here'): 1,
 ('here', 'all'): 1,
 ('all', 'the'): 64,
 ('the', 'carrying'): 1,
 ('carrying', 'on'): 1,
 ('on', 'in'): 5,
 ('in', 'the'): 366,
 ('the', 'neighbors'): 2,
 ('neighbors', 'room'): 1,
 ('room', 'till'): 1,
 ('till', 'am'): 1,
 ('am', 'This'): 1,
 ('This', 'was'): 22,
 ('was', 'very'): 119,
 ('very', 'beautiful'): 2,
 ('beautiful', 'building'): 1,
 ('building', 'the'): 2,
 ('the', 'room'): 209,
 ('room', 'was'): 152,
 ('was', 'a'): 149,
 ('a', 'lot'): 27,
 ('lo

In [142]:
total_bigram_count = sum(train_deceptive_dictionary_pairs.values())
print(total_bigram_count)

76094


In [143]:
train_deceptive_dictionary_bigram = {}
for key,val in train_deceptive_dictionary_pairs.items():
    if key[0] == "START":
        train_deceptive_dictionary_bigram[key] = val/len(train_deceptive)
    else:
        train_deceptive_dictionary_bigram[key] = val/train_deceptive_dictionary[key[0]]

In [144]:
train_deceptive_dictionary_bigram

{('START', 'I'): 0.26171875,
 ('I', 'was'): 0.13078470824949698,
 ('was', 'here'): 0.0014627011214041932,
 ('here', 'on'): 0.02857142857142857,
 ('on', 'business'): 0.0332541567695962,
 ('business', 'so'): 0.03389830508474576,
 ('so', 'I'): 0.15217391304347827,
 ('I', 'needed'): 0.005633802816901409,
 ('needed', 'to'): 0.3225806451612903,
 ('to', 'get'): 0.032034632034632034,
 ('get', 'some'): 0.028409090909090908,
 ('some', 'rest'): 0.009900990099009901,
 ('rest', 'the'): 0.05263157894736842,
 ('the', 'next'): 0.010026901442895574,
 ('next', 'day'): 0.09322033898305085,
 ('day', 'but'): 0.039603960396039604,
 ('but', 'the'): 0.1201923076923077,
 ('the', 'walls'): 0.0034238200048911714,
 ('walls', 'were'): 0.25,
 ('were', 'so'): 0.031413612565445025,
 ('so', 'thin'): 0.013043478260869565,
 ('thin', 'that'): 0.07692307692307693,
 ('that', 'I'): 0.12844036697247707,
 ('I', 'could'): 0.02937625754527163,
 ('could', 'here'): 0.0049504950495049506,
 ('here', 'all'): 0.009523809523809525,
 (

<br>

<br>

<br>

# Perplexity calculation functions:

In [145]:
import math    

In [325]:
def perplexity_unigram_unk(sentence_as_list,unigram_dictionary):
    prob_sum,length_of_sentence = 0,len(sentence_as_list)
    for word in sentence_as_list:
        if word in unigram_dictionary.keys():
            prob_sum += (math.log(unigram_dictionary[word],2))
        else:
            prob_sum += (math.log(unigram_dictionary["UNK"],2))
    l = (prob_sum/length_of_sentence)
    return 2**-l

In [327]:
def perplexity_bigram_unk(sentence_as_list,bigram_dictionary):
    prob_sum,length_of_sentence = 0,len(sentence_as_list)
    for i in range(len(sentence_as_list)-1):
        if (sentence_as_list[i],sentence_as_list[i+1]) in bigram_dictionary.keys():
            prob_sum += (math.log(bigram_dictionary[(sentence_as_list[i],sentence_as_list[i+1])],2))
        else:
            prob_sum += (math.log(bigram_dictionary[("UNK","UNK")],2))
    l = (prob_sum/length_of_sentence)
    return 2**-l

In [437]:
def perplexity_bigram_matrix(sentence_as_list,bigram_matrix):
    prob_sum,length_of_sentence = 0,len(sentence_as_list)
    for i in range(len(sentence_as_list)-1):
        if sentence_as_list[i] in bigram_matrix.index and sentence_as_list[i] != "END" and sentence_as_list[i+1] in bigram_matrix.columns and sentence_as_list[i+1] != "START":
            prob_sum += (math.log(bigram_matrix.at[sentence_as_list[i],sentence_as_list[i+1]],2))
    l = (prob_sum/length_of_sentence)
    return 2**-l


<br>

<br>

# Validation data (Importing)

In [488]:
import pandas as pd

In [489]:
validation_truthful = pd.read_csv("./A1_DATASET/validation/truthful.txt", sep="\n", header=None)
validation_truthful.head()

0
0  I stayed for four nights while attending a con...
1  we love the location and proximity to everythi...
2  I want to issue a travel-warning to folks who ...
3  Just got back from three nights at the Knicker...
4  My family and I stayed at this hotel during an...

In [490]:
validation_deceptive = pd.read_csv("./A1_DATASET/validation/deceptive.txt", sep="\n", header=None)
validation_deceptive.head()

0
0  My stay at the Hotel Allegro Chicago was an am...
1  The Affinia Chicago is a wonderful place to st...
2  I was recently a guest at the Sheraton Chicago...
3  Grant it , this hotel seems very nice , but I ...
4  The Hotel Monaco claims to be a `` boutique Ch...

<br>

<br>

<br>

# Smoothing techniques:
## 1. Add-1 smoothing
## 2. Add-k smoothing (with k=0.5)
## 3. Add-k smoothing (with k=0.001)

<br>

# 1. Add-1 smoothing technique (Laplace)

In [440]:
wordslist_truthful, wordslist_deceptive = [],[]
for key in train_truthful_dictionary.keys():
    wordslist_truthful.append(key)
for key in train_deceptive_dictionary.keys():
    wordslist_deceptive.append(key)

In [441]:
wordslist_truthful.insert(0,"START")
wordslist_truthful.append("END")
wordslist_deceptive.insert(0,"START")
wordslist_deceptive.append("END")

In [442]:
vocab_size_truthful = len(wordslist_truthful)
vocab_size_deceptive = len(wordslist_deceptive)

In [443]:
import numpy as np

In [444]:
row_names_truthful,column_names_truthful = wordslist_truthful,wordslist_truthful
matrix = np.zeros((vocab_size_truthful, vocab_size_truthful))
add_one_matrix_truthful = pd.DataFrame(matrix, columns=column_names_truthful, index=row_names_truthful)

row_names_deceptive,column_names_deceptive = wordslist_deceptive,wordslist_deceptive
matrix = np.zeros((vocab_size_deceptive, vocab_size_deceptive))
add_one_matrix_deceptive = pd.DataFrame(matrix, columns=column_names_deceptive, index=row_names_deceptive)

In [445]:
add_one_matrix_truthful.head()

START    I  booked  two  rooms  four  months   in  advance   at  ...  \
START     0.0  0.0     0.0  0.0    0.0   0.0     0.0  0.0      0.0  0.0  ...   
I         0.0  0.0     0.0  0.0    0.0   0.0     0.0  0.0      0.0  0.0  ...   
booked    0.0  0.0     0.0  0.0    0.0   0.0     0.0  0.0      0.0  0.0  ...   
two       0.0  0.0     0.0  0.0    0.0   0.0     0.0  0.0      0.0  0.0  ...   
rooms     0.0  0.0     0.0  0.0    0.0   0.0     0.0  0.0      0.0  0.0  ...   

        Athletic  lanes  consigned  leftover  stirrers  Yahoo  guarantee  \
START        0.0    0.0        0.0       0.0       0.0    0.0        0.0   
I            0.0    0.0        0.0       0.0       0.0    0.0        0.0   
booked       0.0    0.0        0.0       0.0       0.0    0.0        0.0   
two          0.0    0.0        0.0       0.0       0.0    0.0        0.0   
rooms        0.0    0.0        0.0       0.0       0.0    0.0        0.0   

        beware  advertise  END  
START      0.0        0.0  0.0  
I          0.0        0.0  0.0  
booked     0.0        0.0  0.0  
two        0.0        0.0  0.0  
rooms      0.0        0.0  0.0  

[5 rows x 6507 columns]

In [446]:
add_one_matrix_deceptive.head()

START    I  was  here   on  business   so  needed   to  get  ...  \
START    0.0  0.0  0.0   0.0  0.0       0.0  0.0     0.0  0.0  0.0  ...   
I        0.0  0.0  0.0   0.0  0.0       0.0  0.0     0.0  0.0  0.0  ...   
was      0.0  0.0  0.0   0.0  0.0       0.0  0.0     0.0  0.0  0.0  ...   
here     0.0  0.0  0.0   0.0  0.0       0.0  0.0     0.0  0.0  0.0  ...   
on       0.0  0.0  0.0   0.0  0.0       0.0  0.0     0.0  0.0  0.0  ...   

       recoup  rockin  ROYAL  gifts  tree  informal  tranquility  cleaners  \
START     0.0     0.0    0.0    0.0   0.0       0.0          0.0       0.0   
I         0.0     0.0    0.0    0.0   0.0       0.0          0.0       0.0   
was       0.0     0.0    0.0    0.0   0.0       0.0          0.0       0.0   
here      0.0     0.0    0.0    0.0   0.0       0.0          0.0       0.0   
on        0.0     0.0    0.0    0.0   0.0       0.0          0.0       0.0   

       beutiful  END  
START       0.0  0.0  
I           0.0  0.0  
was         0.0  0.0  
here        0.0  0.0  
on          0.0  0.0  

[5 rows x 5563 columns]

In [447]:
train_truthful_dictionary_pairs

{('START', 'I'): 109,
 ('I', 'booked'): 20,
 ('booked', 'two'): 1,
 ('two', 'rooms'): 3,
 ('rooms', 'four'): 1,
 ('four', 'months'): 1,
 ('months', 'in'): 2,
 ('in', 'advance'): 7,
 ('advance', 'at'): 1,
 ('at', 'the'): 329,
 ('the', 'Talbott'): 22,
 ('Talbott', 'We'): 2,
 ('We', 'were'): 62,
 ('were', 'placed'): 1,
 ('placed', 'on'): 2,
 ('on', 'the'): 213,
 ('the', 'top'): 11,
 ('top', 'floor'): 4,
 ('floor', 'next'): 1,
 ('next', 'to'): 25,
 ('to', 'the'): 263,
 ('the', 'elevators'): 10,
 ('elevators', 'which'): 2,
 ('which', 'are'): 3,
 ('are', 'used'): 1,
 ('used', 'all'): 1,
 ('all', 'night'): 13,
 ('night', 'long'): 4,
 ('long', 'When'): 1,
 ('When', 'speaking'): 1,
 ('speaking', 'to'): 1,
 ('the', 'front'): 80,
 ('front', 'desk'): 97,
 ('desk', 'I'): 4,
 ('I', 'was'): 183,
 ('was', 'told'): 20,
 ('told', 'that'): 19,
 ('that', 'they'): 35,
 ('they', 'were'): 39,
 ('were', 'simply'): 1,
 ('simply', 'honoring'): 1,
 ('honoring', 'my'): 1,
 ('my', 'request'): 3,
 ('request', 'for'

In [448]:
for pair in train_truthful_dictionary_pairs.keys():
    if pair[0] == "START":
        add_one_matrix_truthful.at[pair[0],pair[1]] = (train_truthful_dictionary_pairs[pair]+1)/(len(train_truthful)+vocab_size_truthful)
    else:
        add_one_matrix_truthful.at[pair[0],pair[1]] = (train_truthful_dictionary_pairs[pair]+1)/(train_truthful_dictionary[pair[0]]+vocab_size_truthful)
    

In [449]:
for pair in train_deceptive_dictionary_pairs.keys():
    if pair[0] == "START":
        add_one_matrix_deceptive.at[pair[0],pair[1]] = (train_deceptive_dictionary_pairs[pair]+1)/(len(train_deceptive)+vocab_size_deceptive)
    else:
        add_one_matrix_deceptive.at[pair[0],pair[1]] = (train_deceptive_dictionary_pairs[pair]+1)/(train_deceptive_dictionary[pair[0]]+vocab_size_deceptive)
    

In [450]:
add_one_matrix_truthful.head()

START         I    booked       two     rooms      four  months  \
START     0.0  0.015672  0.000000  0.000000  0.000000  0.000000     0.0   
I         0.0  0.000000  0.002572  0.000000  0.000000  0.000000     0.0   
booked    0.0  0.000304  0.000000  0.000304  0.000000  0.000000     0.0   
two       0.0  0.000000  0.000000  0.000000  0.000604  0.000000     0.0   
rooms     0.0  0.000598  0.000000  0.000000  0.000000  0.000299     0.0   

              in  advance        at  ...  Athletic  lanes  consigned  \
START   0.000285      0.0  0.000285  ...       0.0    0.0        0.0   
I       0.000490      0.0  0.000000  ...       0.0    0.0        0.0   
booked  0.000000      0.0  0.000000  ...       0.0    0.0        0.0   
two     0.000000      0.0  0.000000  ...       0.0    0.0        0.0   
rooms   0.000299      0.0  0.000299  ...       0.0    0.0        0.0   

        leftover  stirrers  Yahoo  guarantee  beware  advertise       END  
START        0.0       0.0    0.0        0.0     0.0        0.0  0.000000  
I            0.0       0.0    0.0        0.0     0.0        0.0  0.000000  
booked       0.0       0.0    0.0        0.0     0.0        0.0  0.000000  
two          0.0       0.0    0.0        0.0     0.0        0.0  0.000302  
rooms        0.0       0.0    0.0        0.0     0.0        0.0  0.000000  

[5 rows x 6507 columns]

In [451]:
add_one_matrix_deceptive.head()

START         I       was      here        on  business        so  \
START    0.0  0.022222  0.000000  0.000000  0.000000  0.000000  0.000000   
I        0.0  0.000249  0.040507  0.000000  0.000000  0.000000  0.000249   
was      0.0  0.001051  0.000000  0.000525  0.002495  0.000000  0.005122   
here     0.0  0.000882  0.000706  0.000000  0.000706  0.000000  0.000000   
on       0.0  0.000000  0.000000  0.000000  0.000000  0.002507  0.000000   

         needed        to       get  ...  recoup  rockin  ROYAL  gifts  tree  \
START  0.000000  0.000000  0.000000  ...     0.0     0.0    0.0    0.0   0.0   
I      0.001864  0.000621  0.001491  ...     0.0     0.0    0.0    0.0   0.0   
was    0.000000  0.002101  0.000000  ...     0.0     0.0    0.0    0.0   0.0   
here   0.000000  0.000353  0.000000  ...     0.0     0.0    0.0    0.0   0.0   
on     0.000000  0.000334  0.000000  ...     0.0     0.0    0.0    0.0   0.0   

       informal  tranquility  cleaners  beutiful       END  
START       0.0          0.0       0.0       0.0  0.000000  
I           0.0          0.0       0.0       0.0  0.000000  
was         0.0          0.0       0.0       0.0  0.000263  
here        0.0          0.0       0.0       0.0  0.000882  
on          0.0          0.0       0.0       0.0  0.000000  

[5 rows x 5563 columns]

In [452]:
for i in add_one_matrix_truthful.index:
    for j in add_one_matrix_truthful.columns:
        if add_one_matrix_truthful.at[i,j] == 0 and i!="END" and j!="START":
            if i == "START":
                add_one_matrix_truthful.at[i,j] = 1/(len(train_truthful)+vocab_size_truthful)
            else:
                add_one_matrix_truthful.at[i,j] = 1/(train_truthful_dictionary[i]+vocab_size_truthful)

In [453]:
add_one_matrix_truthful.head()

START         I    booked       two     rooms      four    months  \
START     0.0  0.015672  0.000142  0.000142  0.000142  0.000142  0.000142   
I         0.0  0.000122  0.002572  0.000122  0.000122  0.000122  0.000122   
booked    0.0  0.000304  0.000152  0.000304  0.000152  0.000152  0.000152   
two       0.0  0.000151  0.000151  0.000151  0.000604  0.000151  0.000151   
rooms     0.0  0.000598  0.000149  0.000149  0.000149  0.000299  0.000149   

              in   advance        at  ...  Athletic     lanes  consigned  \
START   0.000285  0.000142  0.000285  ...  0.000142  0.000142   0.000142   
I       0.000490  0.000122  0.000122  ...  0.000122  0.000122   0.000122   
booked  0.000152  0.000152  0.000152  ...  0.000152  0.000152   0.000152   
two     0.000151  0.000151  0.000151  ...  0.000151  0.000151   0.000151   
rooms   0.000299  0.000149  0.000299  ...  0.000149  0.000149   0.000149   

        leftover  stirrers     Yahoo  guarantee    beware  advertise       END  
START   0.000142  0.000142  0.000142   0.000142  0.000142   0.000142  0.000142  
I       0.000122  0.000122  0.000122   0.000122  0.000122   0.000122  0.000122  
booked  0.000152  0.000152  0.000152   0.000152  0.000152   0.000152  0.000152  
two     0.000151  0.000151  0.000151   0.000151  0.000151   0.000151  0.000302  
rooms   0.000149  0.000149  0.000149   0.000149  0.000149   0.000149  0.000149  

[5 rows x 6507 columns]

In [454]:
for i in add_one_matrix_deceptive.index:
    for j in add_one_matrix_deceptive.columns:
        if add_one_matrix_deceptive.at[i,j] == 0 and i!="END" and j!="START":
            if i == "START":
                add_one_matrix_deceptive.at[i,j] = 1/(len(train_deceptive)+vocab_size_deceptive)
            else:
                add_one_matrix_deceptive.at[i,j] = 1/(train_deceptive_dictionary[i]+vocab_size_deceptive)

In [455]:
add_one_matrix_deceptive

START         I       was      here        on  business  \
START          0.0  0.022222  0.000165  0.000165  0.000165  0.000165   
I              0.0  0.000249  0.040507  0.000124  0.000124  0.000124   
was            0.0  0.001051  0.000131  0.000525  0.002495  0.000131   
here           0.0  0.000882  0.000706  0.000176  0.000706  0.000176   
on             0.0  0.000167  0.000167  0.000167  0.000167  0.002507   
...            ...       ...       ...       ...       ...       ...   
informal       0.0  0.000180  0.000180  0.000180  0.000180  0.000180   
tranquility    0.0  0.000180  0.000180  0.000180  0.000180  0.000180   
cleaners       0.0  0.000180  0.000180  0.000180  0.000180  0.000180   
beutiful       0.0  0.000180  0.000180  0.000180  0.000180  0.000180   
END            0.0  0.000000  0.000000  0.000000  0.000000  0.000000   

                   so    needed        to       get  ...    recoup    rockin  \
START        0.000165  0.000165  0.000165  0.000165  ...  0.000165  0.000165   
I            0.000249  0.001864  0.000621  0.001491  ...  0.000124  0.000124   
was          0.005122  0.000131  0.002101  0.000131  ...  0.000131  0.000131   
here         0.000176  0.000176  0.000353  0.000176  ...  0.000176  0.000176   
on           0.000167  0.000167  0.000334  0.000167  ...  0.000167  0.000167   
...               ...       ...       ...       ...  ...       ...       ...   
informal     0.000180  0.000180  0.000180  0.000180  ...  0.000180  0.000180   
tranquility  0.000180  0.000180  0.000180  0.000180  ...  0.000180  0.000180   
cleaners     0.000180  0.000180  0.000180  0.000180  ...  0.000180  0.000180   
beutiful     0.000180  0.000180  0.000180  0.000180  ...  0.000180  0.000180   
END          0.000000  0.000000  0.000000  0.000000  ...  0.000000  0.000000   

                ROYAL     gifts      tree  informal  tranquility  cleaners  \
START        0.000165  0.000165  0.000165  0.000165     0.000165  0.000165   
I            0.000124  0.000124  0.000124  0.000124     0.000124  0.000124   
was          0.000131  0.000131  0.000131  0.000131     0.000131  0.000131   
here         0.000176  0.000176  0.000176  0.000176     0.000176  0.000176   
on           0.000167  0.000167  0.000167  0.000167     0.000167  0.000167   
...               ...       ...       ...       ...          ...       ...   
informal     0.000180  0.000180  0.000180  0.000180     0.000180  0.000180   
tranquility  0.000180  0.000180  0.000180  0.000180     0.000180  0.000180   
cleaners     0.000180  0.000180  0.000180  0.000180     0.000180  0.000180   
beutiful     0.000180  0.000180  0.000180  0.000180     0.000180  0.000180   
END          0.000000  0.000000  0.000000  0.000000     0.000000  0.000000   

             beutiful       END  
START        0.000165  0.000165  
I            0.000124  0.000124  
was          0.000131  0.000263  
here         0.000176  0.000882  
on           0.000167  0.000167  
...               ...       ...  
informal     0.000180  0.000180  
tranquility  0.000180  0.000180  
cleaners     0.000180  0.000180  
beutiful     0.000180  0.000359  
END          0.000000  0.000000  

[5563 rows x 5563 columns]

## Testing the Add-1 smoothing model on the validation set (Truthful)

In [467]:
truth_perplexity_bigram = 0
for row in range(len(validation_truthful)):
    row_as_list = validation_truthful[0][row].split()
    row_as_list = [values for values in row_as_list if values.isalpha()]
    truth_perplexity_bigram += perplexity_bigram_matrix(row_as_list,add_one_matrix_truthful)

print("Total perplexity of the deceptive validation dataset using the Add-1 smoothing model is: " + str(truth_perplexity_bigram))


Total perplexity of the deceptive validation dataset using the Add-1 smoothing model is: 55573.0397336743


## Testing the Add-1 smoothing model on the validation set (Deceptive)

In [466]:
deception_perplexity_bigram = 0
for row in range(len(validation_deceptive)):
    row_as_list = validation_deceptive[0][row].split()
    row_as_list = [values for values in row_as_list if values.isalpha()]
    deception_perplexity_bigram += perplexity_bigram_matrix(row_as_list,add_one_matrix_deceptive)

print("Total perplexity of the deceptive validation dataset using the Add-1 smoothing model is: " + str(deception_perplexity_bigram))

Total perplexity of the deceptive validation dataset using the Add-1 smoothing model is: 40592.94770199786


<br>

<br>

<br>

# 2. Add-k smoothing technique (k=0.5)

In [476]:
row_names_truthful,column_names_truthful = wordslist_truthful,wordslist_truthful
matrix = np.zeros((vocab_size_truthful, vocab_size_truthful))
add_k_matrix_truthful = pd.DataFrame(matrix, columns=column_names_truthful, index=row_names_truthful)

row_names_deceptive,column_names_deceptive = wordslist_deceptive,wordslist_deceptive
matrix = np.zeros((vocab_size_deceptive, vocab_size_deceptive))
add_k_matrix_deceptive = pd.DataFrame(matrix, columns=column_names_deceptive, index=row_names_deceptive)



In [478]:
for pair in train_truthful_dictionary_pairs.keys():
    if pair[0] == "START":
        add_k_matrix_truthful.at[pair[0],pair[1]] = (train_truthful_dictionary_pairs[pair]+0.5)/(len(train_truthful)+(0.5*vocab_size_truthful))
    else:
        add_k_matrix_truthful.at[pair[0],pair[1]] = (train_truthful_dictionary_pairs[pair]+0.5)/(train_truthful_dictionary[pair[0]]+(0.5*vocab_size_truthful))
    

for pair in train_deceptive_dictionary_pairs.keys():
    if pair[0] == "START":
        add_k_matrix_deceptive.at[pair[0],pair[1]] = (train_deceptive_dictionary_pairs[pair]+0.5)/(len(train_deceptive)+(0.5*vocab_size_deceptive))
    else:
        add_k_matrix_deceptive.at[pair[0],pair[1]] = (train_deceptive_dictionary_pairs[pair]+0.5)/(train_deceptive_dictionary[pair[0]]+(0.5*vocab_size_deceptive))
  

In [480]:
add_k_matrix_truthful.head()

START         I    booked      two     rooms      four  months  \
START     0.0  0.029080  0.000000  0.00000  0.000000  0.000000     0.0   
I         0.0  0.000000  0.004173  0.00000  0.000000  0.000000     0.0   
booked    0.0  0.000450  0.000000  0.00045  0.000000  0.000000     0.0   
two       0.0  0.000000  0.000000  0.00000  0.001037  0.000000     0.0   
rooms     0.0  0.001018  0.000000  0.00000  0.000000  0.000436     0.0   

              in  advance        at  ...  Athletic  lanes  consigned  \
START   0.000398      0.0  0.000398  ...       0.0    0.0        0.0   
I       0.000712      0.0  0.000000  ...       0.0    0.0        0.0   
booked  0.000000      0.0  0.000000  ...       0.0    0.0        0.0   
two     0.000000      0.0  0.000000  ...       0.0    0.0        0.0   
rooms   0.000436      0.0  0.000436  ...       0.0    0.0        0.0   

        leftover  stirrers  Yahoo  guarantee  beware  advertise       END  
START        0.0       0.0    0.0        0.0     0.0        0.0  0.000000  
I            0.0       0.0    0.0        0.0     0.0        0.0  0.000000  
booked       0.0       0.0    0.0        0.0     0.0        0.0  0.000000  
two          0.0       0.0    0.0        0.0     0.0        0.0  0.000445  
rooms        0.0       0.0    0.0        0.0     0.0        0.0  0.000000  

[5 rows x 6507 columns]

In [481]:
add_k_matrix_deceptive.head()

START         I       was      here        on  business        so  \
START    0.0  0.040838  0.000000  0.000000  0.000000  0.000000  0.000000   
I        0.0  0.000285  0.061806  0.000000  0.000000  0.000000  0.000285   
was      0.0  0.001552  0.000000  0.000724  0.003828  0.000000  0.007967   
here     0.0  0.001559  0.001213  0.000000  0.001213  0.000000  0.000000   
on       0.0  0.000000  0.000000  0.000000  0.000000  0.004528  0.000000   

         needed        to       get  ...  recoup  rockin  ROYAL  gifts  tree  \
START  0.000000  0.000000  0.000000  ...     0.0     0.0    0.0    0.0   0.0   
I      0.002753  0.000854  0.002184  ...     0.0     0.0    0.0    0.0   0.0   
was    0.000000  0.003207  0.000000  ...     0.0     0.0    0.0    0.0   0.0   
here   0.000000  0.000520  0.000000  ...     0.0     0.0    0.0    0.0   0.0   
on     0.000000  0.000468  0.000000  ...     0.0     0.0    0.0    0.0   0.0   

       informal  tranquility  cleaners  beutiful       END  
START       0.0          0.0       0.0       0.0  0.000000  
I           0.0          0.0       0.0       0.0  0.000000  
was         0.0          0.0       0.0       0.0  0.000310  
here        0.0          0.0       0.0       0.0  0.001559  
on          0.0          0.0       0.0       0.0  0.000000  

[5 rows x 5563 columns]

In [483]:
for i in add_k_matrix_truthful.index:
    for j in add_k_matrix_truthful.columns:
        if add_k_matrix_truthful.at[i,j] == 0 and i!="END" and j!="START":
            if i == "START":
                add_k_matrix_truthful.at[i,j] = 0.5/(len(train_truthful)+(0.5*vocab_size_truthful))
            else:
                add_k_matrix_truthful.at[i,j] = 0.5/(train_truthful_dictionary[i]+(0.5*vocab_size_truthful))

add_k_matrix_truthful.head()

START         I    booked       two     rooms      four    months  \
START     0.0  0.029080  0.000133  0.000133  0.000133  0.000133  0.000133   
I         0.0  0.000102  0.004173  0.000102  0.000102  0.000102  0.000102   
booked    0.0  0.000450  0.000150  0.000450  0.000150  0.000150  0.000150   
two       0.0  0.000148  0.000148  0.000148  0.001037  0.000148  0.000148   
rooms     0.0  0.001018  0.000145  0.000145  0.000145  0.000436  0.000145   

              in   advance        at  ...  Athletic     lanes  consigned  \
START   0.000398  0.000133  0.000398  ...  0.000133  0.000133   0.000133   
I       0.000712  0.000102  0.000102  ...  0.000102  0.000102   0.000102   
booked  0.000150  0.000150  0.000150  ...  0.000150  0.000150   0.000150   
two     0.000148  0.000148  0.000148  ...  0.000148  0.000148   0.000148   
rooms   0.000436  0.000145  0.000436  ...  0.000145  0.000145   0.000145   

        leftover  stirrers     Yahoo  guarantee    beware  advertise       END  
START   0.000133  0.000133  0.000133   0.000133  0.000133   0.000133  0.000133  
I       0.000102  0.000102  0.000102   0.000102  0.000102   0.000102  0.000102  
booked  0.000150  0.000150  0.000150   0.000150  0.000150   0.000150  0.000150  
two     0.000148  0.000148  0.000148   0.000148  0.000148   0.000148  0.000445  
rooms   0.000145  0.000145  0.000145   0.000145  0.000145   0.000145  0.000145  

[5 rows x 6507 columns]

In [484]:
for i in add_k_matrix_deceptive.index:
    for j in add_k_matrix_deceptive.columns:
        if add_k_matrix_deceptive.at[i,j] == 0 and i!="END" and j!="START":
            if i == "START":
                add_k_matrix_deceptive.at[i,j] = 0.5/(len(train_deceptive)+(0.5*vocab_size_deceptive))
            else:
                add_k_matrix_deceptive.at[i,j] = 0.5/(train_deceptive_dictionary[i]+(0.5*vocab_size_deceptive))

add_k_matrix_deceptive.head()

START         I       was      here        on  business        so  \
START    0.0  0.040838  0.000152  0.000152  0.000152  0.000152  0.000152   
I        0.0  0.000285  0.061806  0.000095  0.000095  0.000095  0.000285   
was      0.0  0.001552  0.000103  0.000724  0.003828  0.000103  0.007967   
here     0.0  0.001559  0.001213  0.000173  0.001213  0.000173  0.000173   
on       0.0  0.000156  0.000156  0.000156  0.000156  0.004528  0.000156   

         needed        to       get  ...    recoup    rockin     ROYAL  \
START  0.000152  0.000152  0.000152  ...  0.000152  0.000152  0.000152   
I      0.002753  0.000854  0.002184  ...  0.000095  0.000095  0.000095   
was    0.000103  0.003207  0.000103  ...  0.000103  0.000103  0.000103   
here   0.000173  0.000520  0.000173  ...  0.000173  0.000173  0.000173   
on     0.000156  0.000468  0.000156  ...  0.000156  0.000156  0.000156   

          gifts      tree  informal  tranquility  cleaners  beutiful       END  
START  0.000152  0.000152  0.000152     0.000152  0.000152  0.000152  0.000152  
I      0.000095  0.000095  0.000095     0.000095  0.000095  0.000095  0.000095  
was    0.000103  0.000103  0.000103     0.000103  0.000103  0.000103  0.000310  
here   0.000173  0.000173  0.000173     0.000173  0.000173  0.000173  0.001559  
on     0.000156  0.000156  0.000156     0.000156  0.000156  0.000156  0.000156  

[5 rows x 5563 columns]

## Testing the Add-k smoothing model on the validation set with k=0.5 (Truthful)

In [487]:
truth_perplexity_bigram = 0
for row in range(len(validation_truthful)):
    row_as_list = validation_truthful[0][row].split()
    row_as_list = [values for values in row_as_list if values.isalpha()]
    truth_perplexity_bigram += perplexity_bigram_matrix(row_as_list,add_k_matrix_truthful)

print("Total perplexity of the truthful validation dataset using the Add-0.5 smoothing model is: " + str(truth_perplexity_bigram))


Total perplexity of the truthful validation dataset using the Add-0.5 smoothing model is: 40990.229312883275


## Testing the Add-k smoothing model on the validation set with k=0.5 (Deceptive)

In [486]:
deception_perplexity_bigram = 0
for row in range(len(validation_deceptive)):
    row_as_list = validation_deceptive[0][row].split()
    row_as_list = [values for values in row_as_list if values.isalpha()]
    deception_perplexity_bigram += perplexity_bigram_matrix(row_as_list,add_k_matrix_deceptive)

print("Total perplexity of the deceptive validation dataset using the Add-0.5 smoothing model is: " + str(deception_perplexity_bigram))


Total perplexity of the deceptive validation dataset using the Add-0.5 smoothing model is: 29524.821412595757


<br>

<br>

<br>

<br>

# 3. Add-k smoothing technique (k=0.001)

In [491]:
row_names_truthful,column_names_truthful = wordslist_truthful,wordslist_truthful
matrix = np.zeros((vocab_size_truthful, vocab_size_truthful))
add_k_matrix_truthful = pd.DataFrame(matrix, columns=column_names_truthful, index=row_names_truthful)

row_names_deceptive,column_names_deceptive = wordslist_deceptive,wordslist_deceptive
matrix = np.zeros((vocab_size_deceptive, vocab_size_deceptive))
add_k_matrix_deceptive = pd.DataFrame(matrix, columns=column_names_deceptive, index=row_names_deceptive)


In [504]:
for pair in train_truthful_dictionary_pairs.keys():
    if pair[0] == "START":
        add_k_matrix_truthful.at[pair[0],pair[1]] = (train_truthful_dictionary_pairs[pair]+0.001)/(len(train_truthful)+(0.001*vocab_size_truthful))
    else:
        add_k_matrix_truthful.at[pair[0],pair[1]] = (train_truthful_dictionary_pairs[pair]+0.001)/(train_truthful_dictionary[pair[0]]+(0.001*vocab_size_truthful))
    

for pair in train_deceptive_dictionary_pairs.keys():
    if pair[0] == "START":
        add_k_matrix_deceptive.at[pair[0],pair[1]] = (train_deceptive_dictionary_pairs[pair]+0.001)/(len(train_deceptive)+(0.001*vocab_size_deceptive))
    else:
        add_k_matrix_deceptive.at[pair[0],pair[1]] = (train_deceptive_dictionary_pairs[pair]+0.001)/(train_deceptive_dictionary[pair[0]]+(0.001*vocab_size_deceptive))


In [505]:
for i in add_k_matrix_truthful.index:
    for j in add_k_matrix_truthful.columns:
        if add_k_matrix_truthful.at[i,j] == 0 and i!="END" and j!="START":
            if i == "START":
                add_k_matrix_truthful.at[i,j] = 0.001/(len(train_truthful)+(0.001*vocab_size_truthful))
            else:
                add_k_matrix_truthful.at[i,j] = 0.001/(train_truthful_dictionary[i]+(0.001*vocab_size_truthful))

add_k_matrix_truthful.head()

START             I    booked           two         rooms  \
START     0.0  2.102209e-01  0.000002  1.928614e-06  1.928614e-06   
I         0.0  6.004178e-07  0.012009  6.004178e-07  6.004178e-07   
booked    0.0  1.143908e-02  0.000011  1.143908e-02  1.142766e-05   
two       0.0  7.904701e-06  0.000008  7.904701e-06  2.372201e-02   
rooms     0.0  1.583583e-02  0.000005  5.276850e-06  5.276850e-06   

                four        months        in       advance            at  ...  \
START   1.928614e-06  1.928614e-06  0.001931  1.928614e-06  1.930543e-03  ...   
I       6.004178e-07  6.004178e-07  0.001802  6.004178e-07  6.004178e-07  ...   
booked  1.142766e-05  1.142766e-05  0.000011  1.142766e-05  1.142766e-05  ...   
two     7.904701e-06  7.904701e-06  0.000008  7.904701e-06  7.904701e-06  ...   
rooms   5.282127e-03  5.276850e-06  0.005282  5.276850e-06  5.282127e-03  ...   

            Athletic         lanes     consigned      leftover      stirrers  \
START   1.928614e-06  1.928614e-06  1.928614e-06  1.928614e-06  1.928614e-06   
I       6.004178e-07  6.004178e-07  6.004178e-07  6.004178e-07  6.004178e-07   
booked  1.142766e-05  1.142766e-05  1.142766e-05  1.142766e-05  1.142766e-05   
two     7.904701e-06  7.904701e-06  7.904701e-06  7.904701e-06  7.904701e-06   
rooms   5.276850e-06  5.276850e-06  5.276850e-06  5.276850e-06  5.276850e-06   

               Yahoo     guarantee        beware     advertise           END  
START   1.928614e-06  1.928614e-06  1.928614e-06  1.928614e-06  1.928614e-06  
I       6.004178e-07  6.004178e-07  6.004178e-07  6.004178e-07  6.004178e-07  
booked  1.142766e-05  1.142766e-05  1.142766e-05  1.142766e-05  1.142766e-05  
two     7.904701e-06  7.904701e-06  7.904701e-06  7.904701e-06  7.912606e-03  
rooms   5.276850e-06  5.276850e-06  5.276850e-06  5.276850e-06  5.276850e-06  

[5 rows x 6507 columns]

In [506]:
for i in add_k_matrix_deceptive.index:
    for j in add_k_matrix_deceptive.columns:
        if add_k_matrix_deceptive.at[i,j] == 0 and i!="END" and j!="START":
            if i == "START":
                add_k_matrix_deceptive.at[i,j] = 0.001/(len(train_deceptive)+(0.001*vocab_size_deceptive))
            else:
                add_k_matrix_deceptive.at[i,j] = 0.001/(train_deceptive_dictionary[i]+(0.001*vocab_size_deceptive))

add_k_matrix_deceptive.head()

START         I           was          here            on  \
START    0.0  0.258908  1.932132e-06  1.932132e-06  1.932132e-06   
I        0.0  0.000402  1.304930e-01  4.015156e-07  4.015156e-07   
was      0.0  0.003404  4.862482e-07  1.459231e-03  8.752953e-03   
here     0.0  0.036188  2.714290e-02  9.044617e-06  2.714290e-02   
on       0.0  0.000002  2.344320e-06  2.344320e-06  2.344320e-06   

           business        so        needed        to           get  ...  \
START  1.932132e-06  0.000002  1.932132e-06  0.000002  1.932132e-06  ...   
I      4.015156e-07  0.000402  5.621620e-03  0.001606  4.417074e-03  ...   
was    4.862482e-07  0.018478  4.862482e-07  0.007294  4.862482e-07  ...   
here   9.044617e-06  0.000009  9.044617e-06  0.009054  9.044617e-06  ...   
on     3.282282e-02  0.000002  2.344320e-06  0.002347  2.344320e-06  ...   

             recoup        rockin         ROYAL         gifts          tree  \
START  1.932132e-06  1.932132e-06  1.932132e-06  1.932132e-06  1.932132e-06   
I      4.015156e-07  4.015156e-07  4.015156e-07  4.015156e-07  4.015156e-07   
was    4.862482e-07  4.862482e-07  4.862482e-07  4.862482e-07  4.862482e-07   
here   9.044617e-06  9.044617e-06  9.044617e-06  9.044617e-06  9.044617e-06   
on     2.344320e-06  2.344320e-06  2.344320e-06  2.344320e-06  2.344320e-06   

           informal   tranquility      cleaners      beutiful           END  
START  1.932132e-06  1.932132e-06  1.932132e-06  1.932132e-06  1.932132e-06  
I      4.015156e-07  4.015156e-07  4.015156e-07  4.015156e-07  4.015156e-07  
was    4.862482e-07  4.862482e-07  4.862482e-07  4.862482e-07  4.867344e-04  
here   9.044617e-06  9.044617e-06  9.044617e-06  9.044617e-06  3.618751e-02  
on     2.344320e-06  2.344320e-06  2.344320e-06  2.344320e-06  2.344320e-06  

[5 rows x 5563 columns]

## Testing the Add-k smoothing model on the validation set with k=0.001 (Truthful)

In [507]:
truth_perplexity_bigram = 0
for row in range(len(validation_truthful)):
    row_as_list = validation_truthful[0][row].split()
    row_as_list = [values for values in row_as_list if values.isalpha()]
    truth_perplexity_bigram += perplexity_bigram_matrix(row_as_list,add_k_matrix_truthful)

print("Total perplexity of the truthful validation dataset using the Add-0.001 smoothing model is: " + str(truth_perplexity_bigram))



Total perplexity of the truthful validation dataset using the Add-0.001 smoothing model is: 19030.55422514236


## Testing the Add-k smoothing model on the validation set with k=0.001 (Deceptive)

In [508]:
deception_perplexity_bigram = 0
for row in range(len(validation_deceptive)):
    row_as_list = validation_deceptive[0][row].split()
    row_as_list = [values for values in row_as_list if values.isalpha()]
    deception_perplexity_bigram += perplexity_bigram_matrix(row_as_list,add_k_matrix_deceptive)

print("Total perplexity of the deceptive validation dataset using the Add-0.001 smoothing model is: " + str(deception_perplexity_bigram))


Total perplexity of the deceptive validation dataset using the Add-0.001 smoothing model is: 12837.437817648966


<br>

<br>

## Summary of smoothing results:

<strong>Truthful validation dataset:</strong> <br>
    Total perplexity (Add-1) = 55573.04 <br>
    Total perplexity (Add-0.5) = 40990.23 <br>
    Total perplexity (Add-0.001) = 19030.55 <br><br>

<strong>Deceptive validation dataset:</strong> <br>
    Total perplexity (Add-1) = 40592.95 <br>
    Total perplexity (Add-0.5) = 29524.82 <br>
    Total perplexity (Add-0.001) = 12837.44 <br><br>

<strong>Conclusion: </strong><br>
From the above results, due to <strong>lower perplexity values </strong>, we infer that <strong>Add-k smoothing with k=0.001 </strong>is giving better results when compared to Add-1 and Add-0.5 smoothing techniques.

<br>

<br>

# Unknown Words Handling (UNK)
## After some frequency analysis, we are consolidating every word that occurs less than 45 times in the document, and marking them as "UNK"

## Truthful unigram with UNK

In [1552]:
train_truthful_dictionary_filtered = {k: v for k, v in sorted(train_truthful_dictionary.items(), key=lambda item: item[1], reverse=True)}


In [1553]:
train_truthful_dictionary_filtered = {k: v for k,v in train_truthful_dictionary_filtered.items() if v>=45}

In [1554]:
len(train_truthful_dictionary_filtered)

226

In [1555]:
train_truthful_dictionary_filtered["UNK"] = sum(train_truthful_dictionary.values())-sum(train_truthful_dictionary_filtered.values())


In [1556]:
train_truthful_dictionary_filtered["UNK"]

25524

In [1557]:
train_truthful_dictionary_filtered_unigram = {key:val/sum(train_truthful_dictionary_filtered.values()) for key,val in train_truthful_dictionary_filtered.items()}


In [1558]:
train_truthful_dictionary_filtered_unigram

{'the': 0.05508891999792606,
 'and': 0.03307927619640172,
 'a': 0.028711048893036762,
 'to': 0.026922279255457045,
 'was': 0.023591019857935398,
 'I': 0.021504121947425726,
 'in': 0.015710063773526208,
 'of': 0.013480582775963083,
 'The': 0.013415772281847876,
 'for': 0.012923212526572303,
 'hotel': 0.011964017213667237,
 'room': 0.011367760667807331,
 'is': 0.010823352517239591,
 'at': 0.009293824856120703,
 'it': 0.009073469176129,
 'that': 0.008580909420853424,
 'we': 0.008490174729092135,
 'on': 0.00790688028205527,
 'with': 0.0076346762067714,
 'were': 0.007492093119717945,
 'not': 0.007388396329133613,
 'had': 0.006805101882096749,
 'We': 0.005975527557422098,
 'this': 0.005871830766837767,
 'but': 0.005845906569191684,
 'have': 0.0055996266915538965,
 'very': 0.005560740395084772,
 'my': 0.005431119406854358,
 'you': 0.005249650023331778,
 'stay': 0.0051200290351013636,
 'our': 0.005107066936278323,
 'they': 0.0049255975527557425,
 'from': 0.004510810390418417,
 'great': 0.00405

<br>

## Truthful bigram with UNK

In [1544]:
train_truthful_dictionary_pairs_filtered = {k:v for k, v in sorted(train_truthful_dictionary_pairs.items(), key=lambda item: item[1], reverse=True)}


In [1545]:
train_truthful_dictionary_pairs_filtered = {k:v for k,v in train_truthful_dictionary_pairs_filtered.items() if v>=45}

In [1546]:
len(train_truthful_dictionary_pairs_filtered)

95

In [1547]:
train_truthful_dictionary_pairs_filtered[("UNK","UNK")] = sum(train_truthful_dictionary_pairs.values())-sum(train_truthful_dictionary_pairs_filtered.values())


In [1548]:
train_truthful_dictionary_pairs_filtered[("UNK","UNK")]

68313

In [1549]:
train_truthful_dictionary_filtered_bigram = {key:val/sum(train_truthful_dictionary_pairs_filtered.values()) for key,val in train_truthful_dictionary_pairs_filtered.items()}


In [1550]:
train_truthful_dictionary_filtered_bigram

{('in', 'the'): 0.0050733968580994075,
 ('of', 'the'): 0.00440381148596446,
 ('at', 'the'): 0.004236415142930724,
 ('the', 'hotel'): 0.0036955961885140358,
 ('and', 'the'): 0.003450939994849343,
 ('to', 'the'): 0.0033865567859902137,
 ('on', 'the'): 0.0027427246973989183,
 ('the', 'room'): 0.0025495750708215297,
 ('I', 'was'): 0.002356425444244141,
 ('for', 'a'): 0.002356425444244141,
 ('room', 'was'): 0.0023049188771568376,
 ('this', 'hotel'): 0.0021632758176667525,
 ('it', 'was'): 0.0019443729075457122,
 ('for', 'the'): 0.0018928663404584084,
 ('was', 'very'): 0.0018799896986865826,
 ('from', 'the'): 0.0017769765645119753,
 ('and', 'I'): 0.0017383466391964976,
 ('was', 'a'): 0.0017254699974246716,
 ('had', 'a'): 0.001661086788565542,
 ('with', 'a'): 0.0016224568632500643,
 ('a', 'great'): 0.001532320370847283,
 ('we', 'were'): 0.0014936904455318052,
 ('in', 'a'): 0.0014936904455318052,
 ('to', 'be'): 0.0014808138037599795,
 ('I', 'had'): 0.0014550605202163277,
 ('is', 'a'): 0.0014550

<br>

## Truthful unigram vs Truthful bigram  

In [1559]:
truth_perplexity_unigram,truth_perplexity_bigram = 0,0
for row in range(len(validation_truthful)):
    row_as_list = validation_truthful[0][row].split()
    row_as_list = [values for values in row_as_list if values.isalpha()]
    truth_perplexity_unigram += perplexity_unigram_unk(row_as_list,train_truthful_dictionary_filtered_unigram)
    truth_perplexity_bigram += perplexity_bigram_unk(row_as_list,train_truthful_dictionary_filtered_bigram)
print("Total perplexity of the truthful validation dataset using the unigram model is: " + str(truth_perplexity_unigram))
print("Total perplexity of the truthful validation dataset using the unigram model is: " + str(truth_perplexity_bigram))


Total perplexity of the truthful validation dataset using the unigram model is: 3199.871430754043
Total perplexity of the truthful validation dataset using the unigram model is: 177.6659867194533


<br>

<br>

## Deceptive unigram with UNK

In [1536]:
train_deceptive_dictionary_filtered = {k:v for k,v in sorted(train_deceptive_dictionary.items(), key=lambda item: item[1], reverse=True)}


In [1537]:
train_deceptive_dictionary_filtered = {k: v for k,v in train_deceptive_dictionary_filtered.items() if v>=45}

In [1538]:
len(train_deceptive_dictionary_filtered)

235

In [1539]:
train_deceptive_dictionary_filtered["UNK"] = sum(train_deceptive_dictionary.values())-sum(train_deceptive_dictionary_filtered.values())


In [1540]:
train_deceptive_dictionary_filtered["UNK"]

22721

In [1541]:
train_deceptive_dictionary_filtered_unigram = {key:val/sum(train_deceptive_dictionary_filtered.values()) for key,val in train_deceptive_dictionary_filtered.items()}


In [1542]:
train_deceptive_dictionary_filtered_unigram

{'the': 0.05410018258315472,
 'and': 0.033367733058135535,
 'I': 0.032878198512873436,
 'to': 0.03056283242041756,
 'was': 0.027136090603582862,
 'a': 0.024423804609563123,
 'in': 0.016313407954274828,
 'of': 0.014712497684633908,
 'The': 0.013640813950411473,
 'room': 0.011934058373686857,
 'hotel': 0.011801751739832235,
 'for': 0.011232833214257363,
 'at': 0.010650684025297029,
 'my': 0.01041253208435871,
 'we': 0.009459924320605434,
 'it': 0.009287925696594427,
 'Chicago': 0.00874546849779048,
 'that': 0.008652853854092243,
 'had': 0.007938398031277289,
 'is': 0.007634092773411659,
 'were': 0.00758117011986981,
 'with': 0.0069460982773676275,
 'not': 0.00690640628721124,
 'stay': 0.0057156465825196475,
 'on': 0.005570109285279564,
 'but': 0.005503955968352253,
 'our': 0.005411341324654018,
 'be': 0.005199650710486624,
 'this': 0.005159958720330237,
 'have': 0.005120266730173851,
 'you': 0.0046439628482972135,
 'very': 0.00449842555105713,
 'they': 0.004392580243973433,
 'would': 0.0

<br>

<br>

## Deceptive bigram with UNK

In [1528]:
train_deceptive_dictionary_pairs_filtered = {k:v for k, v in sorted(train_deceptive_dictionary_pairs.items(), key=lambda item: item[1], reverse=True)}

In [1529]:
train_deceptive_dictionary_pairs_filtered = {k:v for k,v in train_deceptive_dictionary_pairs_filtered.items() if v>=45}

In [1530]:
len(train_deceptive_dictionary_pairs_filtered)

120

In [1531]:
train_deceptive_dictionary_pairs_filtered[("UNK","UNK")] = sum(train_deceptive_dictionary_pairs.values())-sum(train_deceptive_dictionary_pairs_filtered.values())


In [1532]:
train_deceptive_dictionary_pairs_filtered[("UNK","UNK")] 

64814

In [1533]:
train_deceptive_dictionary_filtered_bigram = {key:val/sum(train_deceptive_dictionary_pairs_filtered.values()) for key,val in train_deceptive_dictionary_pairs_filtered.items()}


In [1534]:
train_deceptive_dictionary_filtered_bigram

{('at', 'the'): 0.005335506084579599,
 ('in', 'the'): 0.004809840460483087,
 ('of', 'the'): 0.004547007648434831,
 ('I', 'was'): 0.004271033195784162,
 ('and', 'the'): 0.0037453675716876494,
 ('and', 'I'): 0.0034168265566273293,
 ('to', 'the'): 0.0030094356979525325,
 ('the', 'hotel'): 0.002956869135542881,
 ('the', 'room'): 0.0027466028859042765,
 ('it', 'was'): 0.0026151864798801485,
 ('this', 'hotel'): 0.0025889031986753228,
 ('I', 'had'): 0.002418061870843956,
 ('in', 'Chicago'): 0.002234078902410177,
 ('to', 'be'): 0.0021420874181932873,
 ('I', 'would'): 0.0020763792151812233,
 ('room', 'was'): 0.001997529371566746,
 ('was', 'a'): 0.001958104449759508,
 ('for', 'the'): 0.0019055378873498567,
 ('on', 'the'): 0.001892396246747444,
 ('for', 'a'): 0.0018661129655426184,
 ('we', 'were'): 0.0018135464031329672,
 ('START', 'I'): 0.0017609798407233158,
 ('was', 'very'): 0.0015638552316871238,
 ('stayed', 'at'): 0.0015638552316871238,
 ('to', 'stay'): 0.001550713591084711,
 ('stay', 'at'):

<br>

## Deceptive unigram vs Deceptive bigram  

In [1543]:
deception_perplexity_unigram,deception_perplexity_bigram = 0,0
for row in range(len(validation_truthful)):
    row_as_list = validation_truthful[0][row].split()
    row_as_list = [values for values in row_as_list if values.isalpha()]
    deception_perplexity_unigram += perplexity_unigram_unk(row_as_list,train_deceptive_dictionary_filtered_unigram)
    deception_perplexity_bigram += perplexity_bigram_unk(row_as_list,train_deceptive_dictionary_filtered_bigram)
print("Total perplexity of the deceptive validation dataset using the unigram model is: " + str(deception_perplexity_unigram))
print("Total perplexity of the deceptive validation dataset using the bigram model is: " + str(deception_perplexity_bigram))

Total perplexity of the deceptive validation dataset using the unigram model is: 3084.2949448339464
Total perplexity of the deceptive validation dataset using the bigram model is: 194.79138667436527


<br>

## Summary of unknown words handling results:

### Total perplexities:
<strong>Truthful validation dataset:</strong> <br>
    Unigram UNK model = 3199.87 <br>
    Bigram UNK model = 177.67 <br><br>

<strong>Deceptive validation dataset:</strong> <br>
    Unigram UNK model = 3084.29 <br>
    Bigram UNK model = 194.79 <br><br>

<strong>Conclusion: </strong><br>
From the above results, due to lower perplexity values, we infer that the Bigram UNK models (Truthful and Deceptive) are giving better results when compared to Unigram UNK models (Truthful and Deceptive).

<br>

# Results from validation testing:


<h3>The total perplexities obtained from the various models are ordered as follows: </h3>
<h4>Add-1 smoothing models > Add-0.5 smoothing models > Add-0.001 smoothing models > Unigram UNK models > Bigram UNK models</h4><br>
Hence, from the above computations, we infer that the <strong>Bigram UNK models</strong> produce the <strong>least perplexities</strong> relatively and can be applied on the test dataset.


<br>

<br>

# Final testing with the test dataset using the best model obtained from training and validation steps (Bigram UNK model)

In [502]:
test_data = pd.read_csv("./A1_DATASET/test/test.txt", sep="\n", header=None)
test_data_labels = pd.read_csv("./A1_DATASET/test/test_labels.txt", sep="\n", header=None)

In [1561]:
deception_perplexity_unigram,truth_perplexity_bigram, incorrect = 0,0,0
for row in range(len(test_data)):
    row_as_list = test_data[0][row].split()
    row_as_label = str(test_data_labels[0][row]).split()[0]
    row_as_list = [values for values in row_as_list if values.isalpha()]
    truth_perplexity_bigram = perplexity_bigram_unk(row_as_list,train_truthful_dictionary_filtered_bigram)
    deception_perplexity_bigram = perplexity_bigram_unk(row_as_list,train_deceptive_dictionary_filtered_bigram)
    if(truth_perplexity_bigram < deception_perplexity_bigram and row_as_label == '1'):
        incorrect += 1
    if (deception_perplexity_bigram < truth_perplexity_bigram and row_as_label == '0'):
        incorrect += 1
accuracy_result = (1- (incorrect/len(test_data)))*100
print("The accuracy obtained for the test dataset is: " + str(accuracy_result) + " percent.")

The accuracy obtained for the test dataset is: 44.827586206896555 percent.
